pkg/inspector/image-inspector.go

```diff
@@ -221,7 +183,70 @@ func (i *defaultImageInspector) Inspect() error {
 	return nil
 }
 
-func handleTarStream(reader io.ReadCloser, destination string) {
+// createAndExtractImage creates a docker container based on the option's image with containerName.
+// It will then insepct the container and image and then attempt to extract the image to
+// option's destination path.  If the destination path is empty it will write to a temp directory
+// and update the option's destination path with a /var/tmp directory.  /var/tmp is used to
+// try and ensure it is a non-in-memory tmpfs.
+func (i *defaultImageInspector) createAndExtractImage(client *docker.Client, containerName string) (*docker.Image, error) {
+	container, err := client.CreateContainer(docker.CreateContainerOptions{
+		Name: containerName,
+		Config: &docker.Config{
+			Image: i.opts.Image,
+			// For security purpose we don't define any entrypoint and command
+			Entrypoint: []string{""},
+			Cmd:        []string{""},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create docker container: %v\n", err)
+	}
+
+	// delete the container when we are done extracting it
+	defer func() {
+		client.RemoveContainer(docker.RemoveContainerOptions{
+			ID: container.ID,
+		})
+	}()
+
+	containerMetadata, err := client.InspectContainer(container.ID)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to get docker container information: %v\n", err)
+	}
+
+	imageMetadata, err := client.InspectImage(containerMetadata.Image)
+	if err != nil {
+		return imageMetadata, fmt.Errorf("Unable to get docker image information: %v\n", err)
+	}
+
+	if i.opts.DstPath, err = createOutputDir(i.opts.DstPath, "image-inspector-"); err != nil {
+		return imageMetadata, err
+	}
+
+	reader, writer := io.Pipe()
+	notificationChannel := make(chan int)
+	go handleTarStream(reader, i.opts.DstPath, notificationChannel)
+
+	log.Printf("Extracting image %s to %s", i.opts.Image, i.opts.DstPath)
+	err = client.CopyFromContainer(docker.CopyFromContainerOptions{
+		Container:    container.ID,
+		OutputStream: writer,
+		Resource:     "/",
+	})
+	if err != nil {
+		return imageMetadata, fmt.Errorf("Unable to extract container: %v\n", err)
+	}
+
+	select {
+	case <-notificationChannel:
+		break
+	case <-time.After(time.Minute * 2):
+		return imageMetadata, fmt.Errorf("timeout occured waiting for handleTarStream to finish")
```

Comment:
Why don't we just not use a separate go routine when calling `handleTarStream`?  Then all this goes away.